File: hylian-client/src/main/java/xin/manong/hylian/client/core/HylianClient.java
package xin.manong.hylian.client.core;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.Getter;
import xin.manong.hylian.client.common.Constants;
import xin.manong.hylian.client.config.HylianClientConfig;
import xin.manong.hylian.client.util.SessionUtils;
import xin.manong.hylian.model.Permission;
import xin.manong.hylian.model.Role;
import xin.manong.hylian.model.Tenant;
import xin.manong.hylian.model.User;
import xin.manong.weapon.base.http.HttpRequest;
import xin.manong.weapon.base.http.RequestFormat;
import xin.manong.weapon.jersey.WebResponse;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * hylian客户端
 *
 * @author frankcl
 * @date 2026-01-28 09:29:20
 */
@Getter
public class HylianClient {

    private final HylianClientConfig config;

    public HylianClient(HylianClientConfig config) {
        assert config != null;
        config.check();
        this.config = config;
    }

    /**
     * 强制注销
     *
     * @param httpRequest HTTP请求
     * @param httpResponse HTTP响应
     * @throws IOException I/O异常
     */
    public void forceLogout(HttpServletRequest httpRequest,
                            HttpServletResponse httpResponse) throws IOException {
        httpResponse.sendRedirect(String.format("%s%s?%s=%s&%s=%s", config.serverURL,
                Constants.SERVER_PATH_LOGOUT, Constants.PARAM_APP_ID, config.appId,
                Constants.PARAM_APP_SECRET, config.appSecret));
        SessionManager.invalidate(httpRequest.getSession().getId());
    }

    /**
     * 获取用户角色列表
     *
     * @param user 用户信息
     * @return 角色列表
     */
    public List<Role> getUserRoles(User user) {
        String requestURL = String.format("%s%s", config.serverURL, Constants.SERVER_PATH_GET_APP_USER_ROLES);
        Map<String, Object> paramMap = new HashMap<>();
        paramMap.put(Constants.PARAM_APP_ID, config.appId);
        paramMap.put(Constants.PARAM_APP_SECRET, config.appSecret);
        paramMap.put(Constants.PARAM_USER_ID, user.id);
        HttpRequest httpRequest = HttpRequest.buildGetRequest(requestURL, paramMap);
        List<Role> roles = HTTPExecutor.executeAndUnwrapList(httpRequest, Role.class);
        return roles == null ? new ArrayList<>() : roles;
    }

    /**
     * 获取用户权限列表
     *
     * @param user 用户信息
     * @return 权限列表
     */
    public List<Permission> getUserPermissions(User user) {
        List<Role> roles = getUserRoles(user);
        if (roles == null || roles.isEmpty()) return new ArrayList<>();
        List<String> roleIds = roles.stream().map(role -> role.id).collect(Collectors.toList());
        return getRolePermissions(roleIds);
    }

    /**
     * 获取角色对应权限列表
     *
     * @param roleIds 角色列表
     * @return 权限列表
     */
    public List<Permission> getRolePermissions(List<String> roleIds) {
        String requestURL = String.format("%s%s", config.serverURL,
                Constants.SERVER_PATH_GET_APP_ROLE_PERMISSIONS);
        Map<String, Object> requestBody = new HashMap<>();
        requestBody.put(Constants.PARAM_ROLE_IDS, roleIds);
        requestBody.put(Constants.PARAM_APP_ID, config.appId);
        requestBody.put(Constants.PARAM_APP_SECRET, config.appSecret);
        HttpRequest httpRequest = HttpRequest.buildPostRequest(requestURL, RequestFormat.JSON, requestBody);
        List<Permission> permissions = HTTPExecutor.executeAndUnwrapList(httpRequest, Permission.class);
        return permissions == null ? new ArrayList<>() : permissions;
    }

    /**
     * 获取所有用户信息
     *
     * @return 用户列表
     */
    public List<User> getUsers() {
        String requestURL = String.format("%s%s", config.serverURL,
                Constants.SERVER_PATH_GET_USERS);
        Map<String, Object> requestBody = new HashMap<>();
        requestBody.put(Constants.PARAM_APP_ID, config.appId);
        requestBody.put(Constants.PARAM_APP_SECRET, config.appSecret);
        HttpRequest httpRequest = HttpRequest.buildPostRequest(requestURL, RequestFormat.JSON, requestBody);
        List<User> users = HTTPExecutor.executeAndUnwrapList(httpRequest, User.class);
        return users == null ? new ArrayList<>() : users;
    }

    /**
     * 批量获取用户信息
     *
     * @param userIds 用户ID列表
     * @return 用户信息列表
     */
    public List<User> batchGetUsers(List<String> userIds) {
        String requestURL = String.format("%s%s", config.serverURL,
                Constants.SERVER_PATH_BATCH_GET_USERS);
        Map<String, Object> requestBody = new HashMap<>();
        requestBody.put(Constants.PARAM_APP_ID, config.appId);
        requestBody.put(Constants.PARAM_APP_SECRET, config.appSecret);
        requestBody.put(Constants.PARAM_USER_IDS, userIds);
        HttpRequest httpRequest = HttpRequest.buildPostRequest(requestURL, RequestFormat.JSON, requestBody);
        List<User> users = HTTPExecutor.executeAndUnwrapList(httpRequest, User.class);
        return users == null ? new ArrayList<>() : users;
    }

    /**
     * 根据用户ID获取用户信息
     *
     * @param userId 用户ID
     * @return 成功返回用户信息，否则返回null
     */
    public User getUserById(String userId) {
        String requestURL = String.format("%s%s", config.serverURL,
                Constants.SERVER_PATH_GET_USER_BY_ID);
        Map<String, Object> requestBody = new HashMap<>();
        requestBody.put(Constants.PARAM_APP_ID, config.appId);
        requestBody.put(Constants.PARAM_APP_SECRET, config.appSecret);
        requestBody.put(Constants.PARAM_USER_ID, userId);
        HttpRequest httpRequest = HttpRequest.buildPostRequest(requestURL, RequestFormat.JSON, requestBody);
        return HTTPExecutor.executeAndUnwrap(httpRequest, User.class);
    }

    /**
     * 是否为应用管理员
     *
     * @param user 用户信息
     * @return 应用管理员返回true，否则返回false
     */
    public boolean isAppAdmin(User user) {
        String requestURL = String.format("%s%s", config.serverURL, Constants.SERVER_PATH_IS_APP_ADMIN);
        Map<String, Object> paramMap = new HashMap<>();
        paramMap.put(Constants.PARAM_USER_ID, user.id);
        paramMap.put(Constants.PARAM_APP_ID, config.appId);
        paramMap.put(Constants.PARAM_APP_SECRET, config.appSecret);
        HttpRequest httpRequest = HttpRequest.buildGetRequest(requestURL, paramMap);
        Boolean success = HTTPExecutor.executeAndUnwrap(httpRequest, Boolean.class);
        return success != null && success;
    }

    /**
     * 根据token向服务端获取用户信息
     *
     * @param token 令牌
     * @return 成功返回用户信息，否则返回null
     */
    User getUser(String token) {
        String requestURL = String.format("%s%s", config.serverURL, Constants.SERVER_PATH_GET_USER);
        Map<String, Object> paramMap = buildTokenRequest(token);
        HttpRequest httpRequest = HttpRequest.buildGetRequest(requestURL, paramMap);
        WebResponse<User> response = HTTPExecutor.execute(httpRequest, User.class);
        if (response == null || !response.status) return null;
        return response.data;
    }

    /**
     * 根据token向服务端获取租户信息
     *
     * @param token 令牌
     * @return 成功返回租户信息，否则返回null
     */
    Tenant getTenant(String token) {
        String requestURL = String.format("%s%s", config.serverURL, Constants.SERVER_PATH_GET_TENANT);
        Map<String, Object> paramMap = buildTokenRequest(token);
        HttpRequest httpRequest = HttpRequest.buildGetRequest(requestURL, paramMap);
        WebResponse<Tenant> response = HTTPExecutor.execute(httpRequest, Tenant.class);
        if (response == null || !response.status) return null;
        return response.data;
    }

    /**
     * 刷新token
     *
     * @param token 当前token
     * @return 成功返回新token，否则返回null
     */
    String refreshToken(String token) {
        String requestURL = String.format("%s%s", config.serverURL, Constants.SERVER_PATH_REFRESH_TOKEN);
        Map<String, Object> requestBody = buildTokenRequest(token);
        HttpRequest httpRequest = HttpRequest.buildPostRequest(requestURL, RequestFormat.JSON, requestBody);
        return HTTPExecutor.executeAndUnwrap(httpRequest, String.class);
    }

    /**
     * 根据code获取token
     *
     * @param code 验证code
     * @param httpRequest HTTP请求
     * @return 成功返回token，否则返回null
     */
    String acquireToken(String code, HttpServletRequest httpRequest) {
        String requestURL = String.format("%s%s", config.serverURL, Constants.SERVER_PATH_ACQUIRE_TOKEN);
        Map<String, Object> paramMap = new HashMap<>();
        paramMap.put(Constants.PARAM_CODE, code);
        paramMap.put(Constants.PARAM_APP_ID, config.appId);
        paramMap.put(Constants.PARAM_APP_SECRET, config.appSecret);
        paramMap.put(Constants.PARAM_SESSION_ID, SessionUtils.getSessionID(httpRequest));
        HttpRequest request = HttpRequest.buildGetRequest(requestURL, paramMap);
        return HTTPExecutor.executeAndUnwrap(request, String.class);
    }

    /**
     * 移除会话活动记录
     *
     * @param sessionId 会话ID
     * @return 成功返回true，否则返回false
     */
    boolean removeActivity(String sessionId) {
        String requestURL = String.format("%s%s", config.serverURL, Constants.SERVER_PATH_REMOVE_ACTIVITY);
        Map<String, Object> body = new HashMap<>();
        body.put(Constants.PARAM_SESSION_ID, sessionId);
        body.put(Constants.PARAM_APP_ID, config.appId);
        body.put(Constants.PARAM_APP_SECRET, config.appSecret);
        HttpRequest httpRequest = HttpRequest.buildPostRequest(requestURL, RequestFormat.JSON, body);
        Boolean success = HTTPExecutor.executeAndUnwrap(httpRequest, Boolean.class);
        return success != null && success;
    }

    /**
     * 构建token请求
     *
     * @param token 令牌
     * @return token请求
     */
    private Map<String, Object> buildTokenRequest(String token) {
        Map<String, Object> paramMap = new HashMap<>();
        paramMap.put(Constants.PARAM_TOKEN, token);
        paramMap.put(Constants.PARAM_APP_ID, config.appId);
        paramMap.put(Constants.PARAM_APP_SECRET, config.appSecret);
        return paramMap;
    }
}

File: hylian-client-py/src/hylian_client/sdk/hylian_client.py
"""HylianClient：到 Java hylian-server 的 HTTP 客户端（token 认证模式）。

对齐 Java ``core.HylianClient`` 中 token 模式相关方法，剔除 cookie/session/login/logout。
所有响应都被 :class:`WebResponse` 信封包裹，``status`` 为假时返回 ``None``/空列表。
"""

from __future__ import annotations

from typing import Any
from urllib.parse import urlencode

import httpx

from . import constants as C
from .config import HylianClientConfig
from .models import Permission, Role, User
from .web_response import WebResponse


class HylianClient:
    def __init__(self, config: HylianClientConfig, client: httpx.Client | None = None) -> None:
        config.check()
        self.config = config
        self._client = client or httpx.Client(
            verify=config.verify_tls,
            timeout=config.timeout,
        )

    # ---- 底层请求 ----

    def _url(self, path: str) -> str:
        return f"{self.config.server_url}{path}"

    def _app_params(self) -> dict[str, Any]:
        return {
            C.PARAM_APP_ID: self.config.app_id,
            C.PARAM_APP_SECRET: self.config.app_secret,
        }

    def _unwrap(self, resp: httpx.Response) -> Any:
        # 服务端对「无结果」的接口（如无效 token 的 getUser）会返回 204/空 body，
        # 直接 resp.json() 会抛 JSONDecodeError——空响应一律视为无数据。
        if resp.status_code == 204 or not resp.content:
            return None
        envelope = WebResponse.parse(resp.json())
        return envelope.data if envelope.status else None

    def _get(self, path: str, params: dict[str, Any]) -> Any:
        resp = self._client.get(self._url(path), params=params)
        return self._unwrap(resp)

    def _post(self, path: str, body: dict[str, Any]) -> Any:
        resp = self._client.post(self._url(path), json=body)
        return self._unwrap(resp)

    def logout_redirect_url(self) -> str:
        """服务端 logout 的重定向地址（对齐 Java ``forceLogout``）。

        浏览器被重定向到这里，带着 SSO 域上的 TICKET cookie 命中服务端 logout，
        由服务端完成全局注销并清除 TICKET/TOKEN cookie。
        """
        query = urlencode(self._app_params())
        return f"{self._url(C.SERVER_PATH_LOGOUT)}?{query}"

    def apply_code_url(self, redirect_url: str) -> str:
        """未登录时重定向到服务端 ``applyCode`` 的地址（对齐 Java ``HylianShield`` 的重定向）。

        浏览器带着 SSO 域上的 TICKET cookie 命中 ``applyCode``，服务端签发一次性 code
        并重定向回 ``redirect_url?code=<code>``，客户端再用 code 换 token。
        """
        query = urlencode({**self._app_params(), C.PARAM_REDIRECT_URL: redirect_url})
        return f"{self._url(C.SERVER_PATH_APPLY_CODE)}?{query}"

    # ---- 业务接口 ----

    def acquire_token(self, code: str, session_id: str) -> str | None:
        """用一次性 code 向服务端换取 token（对齐 Java ``acquireToken``）。

        ``session_id`` 让服务端把本地会话与其"应用登录活动(activity)"绑定，
        以便会话销毁时通过 ``remove_activity`` 注销。校验失败返回 ``None``。
        """
        params = {
            C.PARAM_CODE: code,
            C.PARAM_SESSION_ID: session_id,
            **self._app_params(),
        }
        data = self._get(C.SERVER_PATH_ACQUIRE_TOKEN, params)
        return data if data is not None else None

    def remove_activity(self, session_id: str) -> bool:
        """注销会话对应的应用登录活动（对齐 Java ``removeActivity``）。"""
        body = {C.PARAM_SESSION_ID: session_id, **self._app_params()}
        data = self._post(C.SERVER_PATH_REMOVE_ACTIVITY, body)
        return bool(data)

    def get_user(self, token: str) -> User | None:
        """用 token 向服务端换取用户信息，校验失败返回 ``None``。"""
        params = {C.PARAM_TOKEN: token, **self._app_params()}
        data = self._get(C.SERVER_PATH_GET_USER, params)
        return User.model_validate(data) if data is not None else None

    def refresh_token(self, token: str) -> str | None:
        """用旧 token 换取新 token，校验失败返回 ``None``（对齐 Java ``refreshToken``）。"""
        body = {C.PARAM_TOKEN: token, **self._app_params()}
        data = self._post(C.SERVER_PATH_REFRESH_TOKEN, body)
        return data if data is not None else None

    def get_user_by_id(self, user_id: str) -> User | None:
        body = {C.PARAM_USER_ID: user_id, **self._app_params()}
        data = self._post(C.SERVER_PATH_GET_USER_BY_ID, body)
        return User.model_validate(data) if data is not None else None

    def get_users(self) -> list[User]:
        data = self._post(C.SERVER_PATH_GET_USERS, dict(self._app_params()))
        return [User.model_validate(u) for u in (data or [])]

    def batch_get_users(self, user_ids: list[str]) -> list[User]:
        body = {C.PARAM_USER_IDS: user_ids, **self._app_params()}
        data = self._post(C.SERVER_PATH_BATCH_GET_USERS, body)
        return [User.model_validate(u) for u in (data or [])]

    def get_user_roles(self, user: User) -> list[Role]:
        params = {C.PARAM_USER_ID: user.id, **self._app_params()}
        data = self._get(C.SERVER_PATH_GET_APP_USER_ROLES, params)
        return [Role.model_validate(r) for r in (data or [])]

    def get_role_permissions(self, role_ids: list[str]) -> list[Permission]:
        body = {C.PARAM_ROLE_IDS: role_ids, **self._app_params()}
        data = self._post(C.SERVER_PATH_GET_APP_ROLE_PERMISSIONS, body)
        return [Permission.model_validate(p) for p in (data or [])]

    def get_user_permissions(self, user: User) -> list[Permission]:
        roles = self.get_user_roles(user)
        if not roles:
            return []
        role_ids = [r.id for r in roles if r.id]
        if not role_ids:
            return []
        return self.get_role_permissions(role_ids)

    def is_app_admin(self, user: User) -> bool:
        params = {C.PARAM_USER_ID: user.id, **self._app_params()}
        data = self._get(C.SERVER_PATH_IS_APP_ADMIN, params)
        return bool(data)

    def close(self) -> None:
        self._client.close()
